### PR TITLE
[Python BQ] Retry get_table for quota errors

### PR DESCRIPTION
**Pull request by ahmedabu98**
Python SDK's BigQuery get table method is only retried for server and timeout errors. We should extend this to also retry for quota errors (403: `quotaExceeded`, `rateLimitExceeded`).

Also adding some logic to `retry_on_server_errors_timeout_or_quota_issues_filter` to make a best-effort attempt at retrying only for transient reasons. Note that BigQuery also returns error code 403 for non-transient errors like `accessDenied`.

### CODE REVIEW

**Comment by ahmedabu98** [comment]
Run Python 3.8 Postcommit

**Comment by ahmedabu98** [comment]
Run Python 3.10 PostCommit

**Review by Abacn — APPROVED** [review]
LGTM

just a note - python postcommit currently failing bigquery read ITs: #28811

**Comment by kennknowles** [comment]
So obviously I have to ask - can you test it?

**Comment by liferoad** [comment]
You probably need to mock the call like https://github.com/apache/beam/pull/28475/files#diff-a7854eda9da1cdd78ece45e7529978d8296841b84626bee1f925483586b52b14

**Comment by ahmedabu98** [comment]
Yup, added some tests

**Comment by ahmedabu98** [comment]
Run Python 3.8 Postcommit

**Comment by ahmedabu98** [comment]
Run Python 3.10 Postcommit

File: sdks/python/apache_beam/utils/retry.py [hunk 1]
```diff
@@ -57,6 +60,18 @@
 # pylint: enable=wrong-import-order, wrong-import-position
 
 _LOGGER = logging.getLogger(__name__)
+_RETRYABLE_REASONS = [
+    "rateLimitExceeded", "quotaExceeded", "internalError", "backendError"
```
Comment (satybald) [inline sdks/python/apache_beam/utils/retry.py:64]: why do we need to retry on `quotaExceeded`? Once the quota exceeded it needs a manual intervention or back-off of substantial amount of time.

File: sdks/python/apache_beam/utils/retry.py [hunk 2]
```diff
@@ -42,9 +43,11 @@
 try:
   from apitools.base.py.exceptions import HttpError
   from google.api_core.exceptions import GoogleAPICallError
+  from google.api_core import exceptions
```
Comment (satybald) [inline sdks/python/apache_beam/utils/retry.py:46]: to make code more readable, would it be better to use import alias to clearly distinguish between `import requests.exceptions`

File: sdks/python/apache_beam/utils/retry.py [hunk 3]
```diff
@@ -57,6 +60,18 @@
 # pylint: enable=wrong-import-order, wrong-import-position
 
 _LOGGER = logging.getLogger(__name__)
+_RETRYABLE_REASONS = [
+    "rateLimitExceeded", "quotaExceeded", "internalError", "backendError"
+]
+_RETRYABLE_TYPES = (
```
Comment (satybald) [inline sdks/python/apache_beam/utils/retry.py:66]: Should we be DRY here and avoid `if exceptions`? i.e.
```
_RETRYABLE_EXCEPTIONS_FROM_MODULE = [
    exceptions.TooManyRequests,
    exceptions.InternalServerError,
    exceptions.BadGateway,
    exceptions.ServiceUnavailable,
    exceptions.DeadlineExceeded
] if exceptions else []

_RETRYABLE_TYPES = tuple([
    *_RETRYABLE_EXCEPTIONS_FROM_MODULE,
    requests.exceptions.ConnectionError,
    requests.exceptions.Timeout,
    ConnectionError
])
```

File: sdks/python/apache_beam/utils/retry.py [hunk 4]
```diff
@@ -168,15 +183,29 @@ def retry_on_server_errors_and_timeout_filter(exception):
...
     if exception.status_code == 403:
+      try:
+        # attempt to extract the reason and check if it's retryable
+        return exception.content["error"]["errors"][0][
+            "reason"] in _RETRYABLE_REASONS
+      except Exception:
+        _LOGGER.debug(
+            "Could not determine if HttpError is non-transient. Will retry: %s",
+            exception)
       return True
```
Comment (satybald) [inline sdks/python/apache_beam/utils/retry.py:199]: we shouldn't blindly retry on all `403` exceptions. Forbidden exceptions belongs to permanent exceptions. 

Here's a guide from GCP about healthy retry policies:

https://cloud.google.com/storage/docs/retry-strategy

File: sdks/python/apache_beam/utils/retry.py [hunk 5]
```diff
@@ -168,15 +183,29 @@ def retry_on_server_errors_and_timeout_filter(exception):
...
+  403 errors include both transient (accessDenied, billingNotEnabled) and
+  non-transient errors (quotaExceeded, rateLimitExceeded). Only retry transient
+  errors."""
   if HttpError is not None and isinstance(exception, HttpError):
     if exception.status_code == 403:
+      try:
+        # attempt to extract the reason and check if it's retryable
+        return exception.content["error"]["errors"][0][
+            "reason"] in _RETRYABLE_REASONS
+      except Exception:
```
Comment (satybald) [inline sdks/python/apache_beam/utils/retry.py:195]: Instead of generic `Exception` shoud it be `KeyError`?

**Review by satybald — CHANGES_REQUESTED** [review]

File: sdks/python/apache_beam/utils/retry.py (hunk 1 above)
Comment (liferoad) [reply sdks/python/apache_beam/utils/retry.py:64]: For some cases, the quota exceeded error could be transient.

File: sdks/python/apache_beam/utils/retry.py (hunk 3 above)
Comment (ahmedabu98) [reply sdks/python/apache_beam/utils/retry.py:66]: We have tests in this repo that run in environments where these libraries are not available. I'll rewrite it in a more clean way

File: sdks/python/apache_beam/utils/retry.py (hunk 4 above)
Comment (ahmedabu98) [reply sdks/python/apache_beam/utils/retry.py:199]: Thanks for this link, I added some of these status codes as retryable.

Looks like another odd behavior is BigQuery can return [`notImplemented` as 501](https://cloud.google.com/bigquery/docs/error-messages#:~:text=notImplemented,for%20more%20information.) 

> we shouldn't blindly retry on all 403 exceptions. Forbidden exceptions belongs to permanent exceptions.

SGTM

File: sdks/python/apache_beam/utils/retry.py (hunk 1 above)
Comment (ahmedabu98) [reply sdks/python/apache_beam/utils/retry.py:64]: Removed

File: sdks/python/apache_beam/utils/retry.py (hunk 5 above)
Comment (ahmedabu98) [reply sdks/python/apache_beam/utils/retry.py:195]: We can get `KeyError, IndexError, TypeError` from this line. Sure I'll list them out instead

File: sdks/python/apache_beam/utils/retry.py (hunk 3 above)
Comment (ahmedabu98) [reply sdks/python/apache_beam/utils/retry.py:66]: Actually, these aren't needed. There's logic elsewhere in this file that takes care of them

File: sdks/python/apache_beam/utils/retry.py (hunk 2 above)
Comment (ahmedabu98) [reply sdks/python/apache_beam/utils/retry.py]: Removed as this is no longer needed

**Comment by ahmedabu98** [comment]
Thanks for the review @satybald, PTAL

File: sdks/python/apache_beam/utils/retry.py [hunk 6]
```diff
@@ -166,17 +168,36 @@ def retry_on_server_errors_and_timeout_filter(exception):
...
+  billingNotEnabled) and transient errors (rateLimitExceeded).
+  Only retry transient errors."""
   if HttpError is not None and isinstance(exception, HttpError):
-    if exception.status_code == 403:
+    if exception.status_code == 429:
       return True
+    if exception.status_code == 403:
+      try:
+        # attempt to extract the reason and check if it's retryable
+        return exception.content["error"]["errors"][0][
```
Comment (satybald) [inline sdks/python/apache_beam/utils/retry.py:182]: could we run a pipeline to double check that his get parsed correctly? Might enougth to trigger 403- foridden exception to check if this retry works or not

File: sdks/python/apache_beam/utils/retry.py [hunk 7]
```diff
@@ -166,17 +168,36 @@ def retry_on_server_errors_and_timeout_filter(exception):
...
+            "reason"] in _RETRYABLE_REASONS
+      except (KeyError, IndexError, TypeError):
+        _LOGGER.debug(
```
Comment (satybald) [inline sdks/python/apache_beam/utils/retry.py:185]: can we make it info or warning? looks to me an important message to be aware

File: sdks/python/apache_beam/utils/retry.py (hunk 6 above)
Comment (ahmedabu98) [reply sdks/python/apache_beam/utils/retry.py]: Couldn't find a way to reproduce a quota error but I was able to look at the HttpError structure with this:

```python
try:
  table = bq.get_table("google.com:clouddfe", "nonexistent", "nonexistent")
except Exception as e:
  print(type(e))                  # >>> <class 'apitools.base.py.exceptions.HttpNotFoundError'>
  print(e.status_code)              # >>> 404
  print(type(e.content))              # >>> <class 'bytes'>
  content = json.loads(e.content)

  print(content.keys())                   # >>> dict_keys(['error'])
  print(content['error'].keys())             # >>> dict_keys(['code', 'message', 'errors', 'status', 'details'])
  print(content['error']['errors'][0].keys())   # >>> dict_keys(['message', 'domain', 'reason', 'debugInfo'])
  print(content['error']['errors'][0]['reason'])  # >>> notFound
```

**Comment by ahmedabu98** [comment]
@satybald sorry for the delay, PTAL

File: sdks/python/apache_beam/utils/retry.py [hunk 8]
```diff
@@ -166,17 +168,38 @@ def retry_on_server_errors_and_timeout_filter(exception):
...
-    if exception.status_code == 403:
+    if exception.status_code == 429:
       return True
+    if exception.status_code == 403:
+      try:
+        # attempt to extract the reason and check if it's retryable
+        content = exception.content
+        if not isinstance(content, dict):
+          content = json.loads(exception.content)
+        return content["error"]["errors"][0]["reason"] in _RETRYABLE_REASONS
```
Comment (Abacn) [inline sdks/python/apache_beam/utils/retry.py:185]: This is a deeply nested information. Is the response guaranteed (documented) to have this format? In any case, would you mind add a comment for the example error json loads, or is there an example error message could share?

**Comment by ahmedabu98** [comment]
stop reviewer notifications

**Review by Abacn — COMMENTED** [review]
Thanks for adding the test. Just have thing to confirm about the assumption over response

File: sdks/python/apache_beam/utils/retry.py (hunk 8 above)
Comment (ahmedabu98) [reply sdks/python/apache_beam/utils/retry.py:185]: I'm basing this off of the structure of the exception returned from `get_table()`:

```python
bq = bigquery_tools.BigQueryWrapper()
try:
  bq.get_table("google.com:clouddfe", "nonexistent", "nonexistent")
except Exception as e:
  print(type(e))                  # >>> <class 'apitools.base.py.exceptions.HttpNotFoundError'>
  print(e.status_code)              # >>> 404
  print(type(e.content))              # >>> <class 'bytes'>
  content = json.loads(e.content)

  print(content.keys())                   # >>> dict_keys(['error'])
  print(content['error'].keys())             # >>> dict_keys(['code', 'message', 'errors', 'status', 'details'])
  print(content['error']['errors'][0].keys())   # >>> dict_keys(['message', 'domain', 'reason', 'debugInfo'])
  print(content['error']['errors'][0]['reason'])  # >>> notFound
```

Comment (Abacn) [reply sdks/python/apache_beam/utils/retry.py:185]: Thanks for the detail. Anyways this is wrapped in a a fail-safe try clause so LGTM

**Review by Abacn — APPROVED** [review]

**Pull request merged**